ethereum/core/src/integration-test/java/org/hyperledger/besu/ethereum/worldstate/PrunerIntegrationTest.java

```diff
@@ -0,0 +1,243 @@
+import org.hyperledger.besu.ethereum.core.MutableWorldState;
+import org.hyperledger.besu.ethereum.core.TransactionReceipt;
+import org.hyperledger.besu.ethereum.core.WorldState;
+import org.hyperledger.besu.ethereum.rlp.RLP;
+import org.hyperledger.besu.ethereum.storage.keyvalue.WorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.storage.keyvalue.WorldStatePreimageKeyValueStorage;
+import org.hyperledger.besu.ethereum.trie.MerklePatriciaTrie;
+import org.hyperledger.besu.ethereum.trie.StoredMerklePatriciaTrie;
+import org.hyperledger.besu.metrics.noop.NoOpMetricsSystem;
+import org.hyperledger.besu.services.kvstore.InMemoryKeyValueStorage;
+import org.hyperledger.besu.testutil.MockExecutorService;
+import org.hyperledger.besu.util.bytes.Bytes32;
+import org.hyperledger.besu.util.bytes.BytesValue;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.junit.Test;
+
+public class PrunerIntegrationTest {
+
+  private final BlockDataGenerator gen = new BlockDataGenerator();
+  private final NoOpMetricsSystem metricsSystem = new NoOpMetricsSystem();
+  private final Map<BytesValue, byte[]> hashValueStore = new HashMap<>();
+  private final InMemoryKeyValueStorage stateStorage = new TestInMemoryStorage(hashValueStore);
+  private final WorldStateStorage worldStateStorage = new WorldStateKeyValueStorage(stateStorage);
+  private final WorldStateArchive worldStateArchive =
+      new WorldStateArchive(
+          worldStateStorage, new WorldStatePreimageKeyValueStorage(new InMemoryKeyValueStorage()));
+  private final InMemoryKeyValueStorage markStorage = new InMemoryKeyValueStorage();
+  private final Block genesisBlock = gen.genesisBlock();
+  private final MutableBlockchain blockchain = createInMemoryBlockchain(genesisBlock);
+
+  @Test
+  public void pruner_smallState_manyOpsPerTx() throws InterruptedException {
+    testPruner(3, 1, 4, 1000);
+  }
+
+  @Test
+  public void pruner_largeState_fewOpsPerTx() throws InterruptedException {
+    testPruner(2, 5, 6, 5);
+  }
+
+  @Test
+  public void pruner_emptyBlocks() throws InterruptedException {
+    testPruner(5, 0, 5, 10);
+  }
+
+  @Test
+  public void pruner_markChainhead() throws InterruptedException {
+    testPruner(4, 2, 10, 20);
+  }
+
+  private void testPruner(
+      final int numCycles,
+      final int accountsPerBlock,
+      final int numBlocksToKeep,
+      final int opsPerTransaction)
+      throws InterruptedException {
+
+    final var markSweepPruner =
+        new MarkSweepPruner(
+            worldStateStorage, blockchain, markStorage, metricsSystem, opsPerTransaction);
+    final var pruner =
+        new Pruner(
+            markSweepPruner,
+            blockchain,
+            new MockExecutorService(),
+            new PruningConfiguration(1, numBlocksToKeep));
```

Comment:
Shouldn't we parameterize and test various options for `blockConfirmationsBeforeMarking`?  This is where one of the bugs was.

Reply:
Good idea, will add that in